proposals/cordova-platform-add-refactor.md

```diff
@@ -0,0 +1,160 @@
+
+The end functionality should be the same in the ideal case, and the API of the `cordova platform add` command should not change (except the removal of some deprecated flags perhaps).
+
+## Platform Add
+
+I would like to separate the implementation of platform adding into 4 steps:
+
+1. **fetch:** Runs npm to install the platform into ./node_modules/ and add it to package.json as a devDependency, then adds it to the list of platforms in the cordova section of package.json
+
+2. **install:** Uses the platform module from ./node_modules/ to create the platform's native project under ./platform/&lt;platform_name&gt;/
+
+3. **plugin-setup:** Runs any installation/preparation code for plugins that need to be installed in the newly-created platform project.
+
+3. **prepare:** Runs the platform's prepare script to copy preferences and resources from config.xml into the project
+
+Each of these steps should be implemented as a stateless function that can be run individually and returns a Promise, which will allow for much easier unit testing and the ability to reuse these smaller steps in other cordova commands (like the proposed `cordova install` command).
+
+What follows is the rough pseudo-code for what each of these steps should do.
+
+
+### Platform Fetch
+Runs npm to install the platform package into ./node_modules/ and add it to package.json as a devDependency, then adds it to the list of platforms in the cordova section of package.json
+
+**Parameters:**
+* `platform_spec : string` - A platform package specification.
+
+  Examples:
+  * `android` (meaning `cordova-android`)
+  * `cordova-electron` (meaning `cordova-electron`)
+  * `@dpogue/android` (meaning `@dpogue/android`)
+  * `ios@^4.0.0` (meaning `cordova-ios` with version `^4.0.0`)
+  * `windows@1.0.0` (meaning `cordova-windows` with version `1.0.0`)
+  * `android@~6.1.0` (meaning `cordova-android` with version `~6.1.0`)
+
+**Returns:** `Promise<string>` - A promise resolving with the name of the installed platform.
+
+```javascript
+function cordova_platform_fetch(platform_spec) {
+    let platform_module = platform_spec
+
+    // Turn "android" into "cordova-android"
+    // But leave "@dpogue/whatever" as it is
+    if (!platform_spec.startsWith('cordova-') && !platform_spec.startsWith('@')) {
+        platform_module = 'cordova-' + platform_spec
+    }
+
+    // Get the platform name with any version specifiers stripped off
+    let platform_name = platform_spec.slice(0, platform_spec.lastIndexOf('@'))
+
+    let pkgJson = new PackageHelper('package.json')
+
+    // TODO: See if there's an easy way to determine if the platform is already
+    // installed with a matching version and avoid the npm invocation
```

Comment:
cordova-fetch already does this, so you can just call it and use the path to the installed package it returns to determine the package name too.

Reply:
Ah nice!